core/src/main/java/kafka/log/remote/RemoteLogManager.java

```diff
@@ -738,6 +750,23 @@ public void copyLogSegmentsToRemote(UnifiedLog log) throws InterruptedException
                                         isCancelled(), isLeader());
                                 return;
                             }
+
+                            copyQuotaManagerLock.lock();
+                            try {
+                                while (rlmCopyQuotaManager.isQuotaExceeded()) {
```

Comment:
If the copy quota gets breached, all the `kafka-rlm-thread-pool` threads will wait for the quota to be available which might delay the deletion of remote log segments since the same thread does both copy and delete.

Do you plan to split the copy/delete operations in a separate thread pool?

Reply:
Yes, this was planned as part of KIP-950 implementation. We will have a separate change for it.

Reply:
@abhijeetk88 @kamalcph , I'm thinking if we could skip this copy when quota exceeds, and waiting for next RLMTask run? In that way, the implementation will be much simiplier without lock and wait. WDYT?

Reply:
> we could skip this copy when quota exceeds, and waiting for next RLMTask run? 

RLMTask schedule interval is configurable and defaults to 30 seconds. The next run will happen after 30 seconds, the quota might be available before the timeout. If we go with the suggested approach, then we may not be utilizing the quota effectively.

Reply:
Fair enough. Thanks.

Reply:
@abhijeetk88 

One drawback of the current approach is that the thread-pool will be shown as all "busy" in the metrics. Could you confirm whether the threads are "shown as" busy when they are in `await`?

Reply:
Good point! [KAFKA-16887](https://issues.apache.org/jira/browse/KAFKA-16887) is created for this issue. 
I'm also thinking, maybe we could still skip this copy when quota exceeds, and waiting for next RLMTask run. And if it's throttled, we can schedule another task in 1 second delay. That will mitigate the issue @kamalcph mentioned above, also get the same result. WDYT?

Reply:
I agree that the theadpool will look busy when the threads are waiting for the quota to be available. However, we will also have another metric for throttling which can be referred to verify if the copy operations are getting throttled.
If we see **threadpool busy and copy being throttled -> copy quota is fully utilized and there is nothing to do**
However, if **threadpool is busy and no throttles are happening -> there are not enough threads in the pool and threadpool size needs to be increased**.

Skipping the current run (as Kamal pointed out) will result in ineffective use of the quota because the next run of the task happens after 30 sec. Also, this may cause starvation for some topic partitions. There is no fairness. For eg. the copy for one topic partition gets skipped, but for another one is picked immediately if the quota is available. These problems will continue to exist if we schedule the next run with 1 sec delay.

It is also not simple to schedule a new RLMTask with a 1-sec delay (if we skip the current run). The schedule of the tasks is already fixed because we are using a ScheduledTheadPoolExecutor with _scheduleWithFixedDelay_ to schedule the tasks. If we need to make the above change, the task needs to own its own scheduling. i.e after each run, it schedules its next run. This will be a larger change.

Reply:
@abhijeetk88 , thanks for the response. I agree we can keep it as is, and just document it clearly in monitoring section: https://kafka.apache.org/documentation/#monitoring . We can improve it in the future if necessary.